File: AMBB.cpp
#include <Rcpp.h>
using namespace Rcpp;

#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>

#include <string>



std::vector<int> row_seed(const std::vector<std::vector<int>>& mat) {
  std::vector<int> row_sums(mat.size());
  for (int i = 0; i < mat.size(); ++i) {
    row_sums[i] = std::accumulate(mat[i].begin(), mat[i].end(), 0);
  }
  int max_sum = *std::max_element(row_sums.begin(), row_sums.end());
  std::vector<int> result;
  for (int i = 0; i < row_sums.size(); ++i) {
    if (row_sums[i] == max_sum) {
      result.push_back(i);
    }
  }
  return result;
}

std::vector<int> column_seed(const std::vector<std::vector<int>>& mat) {
  std::vector<int> column_sums(mat[0].size());
  for (int j = 0; j < mat[0].size(); ++j) {
    for (int i = 0; i < mat.size(); ++i) {
      column_sums[j] += mat[i][j];
    }
  }
  int max_sum = *std::max_element(column_sums.begin(), column_sums.end());
  std::vector<int> result;
  for (int i = 0; i < column_sums.size(); ++i) {
    if (column_sums[i] == max_sum) {
      result.push_back(i);
    }
  }
  return result;
}


int row_diff(std::vector<int> row, std::vector<int> seed){
  int c = 0;
  for (int i = 0; i < seed.size(); i++){
    if (seed[i] != row[i]){
      c++;
    } 
  }
  return c;
}

std::vector<int> Row_difference(std::vector<std::vector<int>> E, std::vector<int> seed){
  std::vector<int> c(seed.size(), 0);
  for (int i = 0; i < E.size(); i++){
    c[i] = row_diff(E[i], seed);
  }
  return c;
}

std::vector<std::vector<int>> Row_cluster(const std::vector<std::vector<int>>& E, int row_threshold, int i) {
  auto seed = E[Row_seed(E)[i]];
  auto row_DV = Row_difference(E, seed);
  std::vector<int> list;
  for (int i = 0; i < row_DV.size(); i++) {
    if (row_DV[i] >= row_threshold) {
      list.push_back(i);
    }
  }
  if (list.empty()) {
    return E;
  }
  std::vector<std::vector<int>> result;
  for (int i = 0; i < E.size(); i++) {
    if (std::find(list.begin(), list.end(), i) == list.end()) {
      result.push_back(E[i]);
    }
  }
  return result;
}


std::vector<std::vector<double>> Col_cluster(std::vector<std::vector<double>> &E, double col_threshold, int i) {
  std::vector<double> seed;
  for (int j = 0; j < E[i].size(); j++) {
    seed.push_back(E[j][i]);
  }
  
  std::vector<double> col_DV;
  for (int j = 0; j < E.size(); j++) {
    double difference = 0;
    for (int k = 0; k < seed.size(); k++) {
      difference += abs(E[j][k] - seed[k]);
    }
    col_DV.push_back(difference);
  }
  
  std::vector<int> list;
  for (int j = 0; j < col_DV.size(); j++) {
    if (col_DV[j] > col_threshold) {
      list.push_back(j);
    }
  }
  
  std::vector<std::vector<double>> result;
  if (list.empty()) {
    return E;
  } else {
    for (int j = 0; j < E.size(); j++) {
      if (std::find(list.begin(), list.end(), j) == list.end()) {
        result.push_back(E[j]);
      }
    }
    return result;
  }
}

std::vector<std::vector<double>> max_sum_matrix(std::vector<std::vector<std::vector<double>>> &matrix_list) {
  double max_sum = -INFINITY;
  std::vector<std::vector<double>> max_matrix = matrix_list[0];
  
  for (int i = 0; i < matrix_list.size(); i++) {
    double sum = 0;
    for (int j = 0; j < matrix_list[i].size(); j++) {
      for (int k = 0; k < matrix_list[i][j].size(); k++) {
        sum += matrix_list[i][j][k];
      }
    }
    if (sum > max_sum) {
      max_sum = sum;
      max_matrix = matrix_list[i];
    }
  }
  
  return max_matrix;
}

vector<vector<int>> AMBB2(vector<vector<int>> &E, int row_threshold, int col_threshold) {
  vector<vector<int>> row_seeds = Row_seed(E);
  vector<vector<int>> Bicluster;
  for (int i = 0; i < row_seeds.size(); i++) {
    vector<vector<int>> row_matrix = Row_cluster(E, row_threshold, i);
    vector<vector<int>> col_seeds = Row_seed(row_matrix);
    for (int j = 0; j < col_seeds.size(); j++) {
      vector<vector<int>> col_matrix = Col_cluster(row_matrix, col_threshold, j);
      if (!col_matrix.empty()) {
        if (find(col_matrix.begin(), col_matrix.end(), 0) == col_matrix.end()) {
          Bicluster.push_back(col_matrix);
        } else {
          Bicluster.push_back(AMBB2(col_matrix, 1, 1));
        }
      }
    }
  }
  return max_sum_matrix(Bicluster);
}


using namespace std;

vector<vector<int>> permute_rows(vector<vector<int>> &M, int row1, int row2) {
  string rows[M.size()];
  for (int i = 0; i < M.size(); i++) {
    rows[i] = to_string(i);
  }
  swap(rows[row1], rows[row2]);
  swap(M[row1], M[row2]);
  return M;
}

vector<vector<int>> permute_columns(vector<vector<int>> &M, int col1, int col2) {
  for (int i = 0; i < M.size(); i++) {
    swap(M[i][col1], M[i][col2]);
  }
  return M;
}

vector<vector<int>> permute(vector<vector<int>> &M, map<int, string> Bicluster_rows, map<int, string> Bicluster_cols) {
  for (int i = 0; i < Bicluster_rows.size(); i++) {
    int row = stoi(Bicluster_rows[i]);
    M = permute_rows(M, i, row);
  }
  for (int j = 0; j < Bicluster_cols.size(); j++) {
    int col = stoi(Bicluster_cols[j]);
    M = permute_columns(M, j, col);
  }
  return M;
}

File: AMBB.R
# On définit une fonction qui nous retourne les row-seeds pour chaque matrice
Row_seed <- function(mat) {
  row_sums <- rowSums(mat)
  max_sum <- max(row_sums)
  return(which(row_sums == max_sum))
}


#  On définit une fonction qui nous retourne les column-seeds pour chaque matrice
Column_seed <- function(mat) {
  Column_sums <- colSums(mat)
  max_sum <- max(Column_sums)
  return(which(Column_sums == max_sum))
}


# On définit une fonction qui prend en entrées deux lignes et nous retournes le nombre de différences entre eux
row_diff <- function(row, seed){
  c=0
  for (i in 1: length(seed)){
    if (seed[i] != row[i]){
      c = c+1
    } 
  }
  return(c)
}

# On définit la fonction qui nous consruit la matrice de différences pour toutes les lignes selon le seed 

Row_difference <- function(E, seed){
  c = rep(0,length(seed))
  for (i in 1:dim(E)[1]){
    c[i] = row_diff(row = E[i,] , seed)
  }
  return(c)
}

# On définit une cette fonction qui prend en entrée une matrice, un row threshold et un seed, et nous retourne retient que 
# les lignes ayant une differnce-value inférieur au threshold
Row_cluster <- function(E, row_threshold,i){
  seed = E[Row_seed(E)[i],]
  row_DV = Row_difference(E,seed = seed)
  list <- c()
  for(i in 1:length(row_DV)){
    if ( row_DV[i]>= row_threshold){
      list <- c(list,i)
    }
  }
  if (length(list) == 0 ){
    return(E) 
  }
  return(E[-list,])
}
# On définit une cette fonction qui prend en entrée une matrice, un  column-threshold et un col-seed, et nous retourne retient que 
# les colonnes ayant une differnce-value inférieur au threshold
Col_cluster <- function(E, col_threshold,i){
  seed = E[,Row_seed(t(E))[i]]
  #seed = E[,5]
  col_DV = Row_difference(t(E),seed)
  list <-c()
  for(i in 1:length(col_DV)){
    if ( col_DV[i]> col_threshold){
      list <- c(list,i)
    }
  }
  if (length(list) == 0 ){
    return(E) 
  }
  return(E[,-list])
}

# On définit une fonction qui prend une liste de matrice et nous retourne la matrice ayant la somme des valeurs maximale
# dans notre cas on va l'utiliser pour retenir le bicluster ayant le nombre maximal des 1 
max_sum_matrix <- function(matrix_list) {
  max_sum <- -Inf
  max_matrix <- matrix_list[[1]]
  
  for (matrix in matrix_list) {
    sum <- sum(matrix)
    if (sum > max_sum) {
      max_sum <- sum
      max_matrix <- matrix
    }
  }
  
  return(max_matrix)
}


# AMBB2 fonction qui fait le biclustering , elle prend en entrée une matrice binaire, un row-threshold et un col-threshold et retourne
#le bicluster maximale
AMBB2 <- function(E, row_threshold, col_threshold){
  row_seeds = Row_seed(E)# la liste des row-seeds
  Bicluster = list()#on initialise notre liste de bicluster par une liste vide
  #Biclusters = list()
  for (i in 1:length(row_seeds)){ #on parcour les row-seeds 
    row_matrix = Row_cluster(E,row_threshold,i) #on construit le row-cluster selon la valeur du row-threshold
    col_seeds = Row_seed(t(row_matrix))#la liste des col-seeds du le row-cluster
    for (j in 1:length(col_seeds)){#on parcours les col-seeds
      col_matrix = Col_cluster(row_matrix, col_threshold,j) #on construit le Col-cluster selon chaque seed
      if (!is.null(dim(col_matrix))){ # on vérifie si la matrice col-matrixest bien définie
        
        if (!any(0 %in% col_matrix)){ # on vérifie si elle contient encore des 0 
          Bicluster <- append(Bicluster, list(col_matrix)) #si elle ne contient pas des 0 on la rajoute à notre liste de Biclusters
        }
        else{
          Bicluster <- append(Bicluster,list((AMBB2(col_matrix,1,1)))) #sinon on refait le même processus pour la col-matrix
        }
        
      }
      
    }
    
    
    
  }
  return(max_sum_matrix(Bicluster)) #On retourn à la fin le Bicluster ayant le nombre maximal des 1
  
  
  
}

#Pour l'affichage du bicluster on définit la fonction permute

# Cette fonction prend une matrice et deux indice de deux lignes et les permute
permute_rows <- function(M, row1, row2){
  rows <- rownames(M)
  rows[c(row1, row2)] <- rows[c(row2, row1)]
  rownames(M) <- rows
  M[c(row1, row2),] <- M[c(row2, row1),]
  return(M)
}
# Cette fonction prend une matrice et deux indice de deux colonnes et les permute
permute_cols <- function(M2, col1, col2){
  cols <- colnames(M2)
  cols[c(col1, col2)] <- cols[c(col2, col1)]
  colnames(M2) <- cols
  M2[, c(col1, col2)] <- M2[, c(col2, col1)]
  return(M2)
}

#Cette fonction prend en entrée une matrice binaire et le bicluster donée par l'algorithme et réarange les lignes
#et les colonnes pour faire apparaitre le Bicluster 
permute <- function(M,Bicluster){
  r= rownames(Bicluster)
  c= colnames(Bicluster)
  for (i in 1:length(r)){
    M <- permute_rows(M,i,as.numeric(r[i]))
  }
  for (j in 1:length(c)){
    M <- permute_cols(M,j,as.numeric(c[j]))
  }
  
  return(M)
  
}














